Dynamic registration of RPC backends

satgera opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#25734 Dynamic registration of RPC backends**

Allow non-pg rpc backends to be plugged in as a backend.

Differential Revision: [D17183789](https://our.internmc.facebook.com/intern/diff/D17183789/)

satgera committed:
[pytorch] OSS changes to support backend other than pg

[pytorch] OSS changes to support backend other than pg
Allow non-pg rpc backends to be plugged in as a backend.

Differential Revision: [D17183789](https://our.internmc.facebook.com/intern/diff/D17183789/)

pietern commented on test/test_rpc.py:
```diff
@@ -123,35 +125,15 @@ def test_self_add(self):
-            dist.init_model_parallel("".join(["a" for _ in range(500)]))
-
-        dist.join_rpc()
-
```
Is this test no longer valid/needed?

mrshenli replied:
If other backend does not throw the same error, let's make this test specific for process group backend.

satgera replied:
Restored

pietern commented on test/test_rpc.py:
```diff
@@ -182,6 +164,7 @@ def test_scalar_add(self):
         )
         self.assertEqual(ret, (torch.ones(n, n) + n))
 
+
```
Remove.

pietern commented on torch/distributed/__init__.py:
```diff
@@ -23,7 +23,8 @@ def is_available():
 
-        def init_model_parallel(worker_name, rpc_backend=RpcBackend.PROCESS_GROUP):
+        def init_model_parallel(worker_name, rpc_backend=RpcBackend.PROCESS_GROUP,
+                                rank=-1, world_size=-1, store=None):
```
Please put every arg on a new line and add docstrings below.

pietern commented on torch/distributed/rpc.py:
```diff
@@ -23,6 +24,15 @@ def wrapper(*args, **kwargs):
+    global _agent
+    _agent = agent
+
+
```
Are these intended to be a public API? If not, prefix with `_`.

satgera replied:
these are public.

pietern replied:
Can you add a docstring for them? E.g. is this just for overriding, when should you call these, etc.

pietern replied:
These are no longer used it seems?

satgera replied:
Removing.

pietern commented on torch/distributed/rpc.py:
```diff
@@ -51,8 +61,9 @@ class RpcBackend(Enum):
-def _init_rpc(name, backend=RpcBackend.PROCESS_GROUP):
+# TODO: add a context managet to wrap _init_rpc and join_rpc
+def _init_rpc(name, backend=RpcBackend.PROCESS_GROUP, world_size=-1,
+              rank=-1, store=None):
```
Please put each arg on a new line.

pietern commented on torch/distributed/rpc_handler.py:
```diff
@@ -0,0 +1,32 @@
+_rpc_init_handlers = {}
+
+
+def register_rpc_handler(backend_str, handler):
```
Shouldn't this be `register_rpc_backend`?

mrshenli commented on torch/distributed/rpc.py:
```diff
@@ -62,10 +73,19 @@ def _init_rpc(name, backend=RpcBackend.PROCESS_GROUP):
-        from .distributed_c10d import _get_default_group
+        from .distributed_c10d import _get_default_group, init_process_group
+
+        init_process_group(
```
Any reason for putting `init_process_group` in `init_rpc`? Would it be better if we keep them separate as what we did for `DistributedDataParallel`?

mrshenli commented on torch/distributed/rpc_handler.py:
```diff
@@ -0,0 +1,32 @@
+            `init_rpc()` function is called with a backend. This sets the global agent.
+    """
+    global _rpc_init_handlers
+    if backend_str in _rpc_init_handlers:
```
Do we need a lock in this function?

mrshenli commented on torch/distributed/rpc_handler.py:
```diff
@@ -0,0 +1,32 @@
+    Arguments:
+        backend (str): backend to identify the handler.
+        handler (function): Handler that is invoked when the
+            `init_rpc()` function is called with a backend. This sets the global agent.
```
it is now renamed to `_init_rpc()`. Or do you mean `init_model_parallel()`?

mrshenli commented on torch/distributed/rpc_handler.py:
```diff
@@ -0,0 +1,32 @@
+
+
+def is_backend_registered(backend_str):
+    global _rpc_init_handlers
```
Don't need global for read-only access.

mrshenli commented on torch/distributed/rpc.py:
```diff
@@ -62,10 +73,19 @@ def _init_rpc(name, backend=RpcBackend.PROCESS_GROUP):
+    elif is_backend_registered(backend):
+        _agent = registered_init_rpc(backend,
+                                     worker_id=rank,
+                                     worker_names=["worker{}".format(i) for i in range(world_size)],
```
Let's add a `TODO` to collect worker names here.

satgera replied:
solved with URL where one can pass in a store or something to identify the workers.

satgera commented on test/test_rpc.py:
```diff
@@ -77,9 +79,9 @@ def _wrap_with_rpc(func):
-        dist.init_model_parallel('worker%d' % self.rank)
+        dist.init_model_parallel('worker%d' % self.rank, rpc_backend=BACKEND,
+                                 rank=self.rank, world_size=self.world_size,
+                                 store=store)
```
Had a discussion with @mrshenli and want to document it here to solicit more feedback.

This diff changes init_model_parallel behaviour and calls init_process_group internally. Since all backends may not support pg like collectives this may make it inconvenient for the user. Also this makes the behavior different than data parallel case where init_process_group should have been initialized before DistributedDataParallel. Another argument against doing this is that each rpc_backend may need different arguments and hence we would need to keep the arguments as a supper set of the all backends.

Based on the above I plan to keep init_process_group to be called if needed by the user outside init_model_parallel and make the arguments to init_model_parallel as a url which can be interpreted by the backend implementation.

satgera committed:
Update on "Dynamic registration of RPC backends"


Allow non-pg rpc backends to be plugged in as a backend.

Differential Revision: [D17183789](https://our.internmc.facebook.com/intern/diff/D17183789/)

mrshenli commented on torch/distributed/rpc.py:
```diff
@@ -61,11 +74,18 @@ def _init_rpc(name, backend=RpcBackend.PROCESS_GROUP):
         group = _get_default_group()
         # TODO: add try-except and destroy _agent in all processes if any fails.
-        _agent = ProcessGroupAgent(name, group)
+        _agent = ProcessGroupAgent(self_name, group)
```
Do we want to check and throws an error/warn if `self_rank != group.rank()` here?

mrshenli commented on torch/distributed/__init__.py:
```diff
@@ -23,7 +23,10 @@ def is_available():
 
-        def init_model_parallel(worker_name, rpc_backend=RpcBackend.PROCESS_GROUP):
+        def init_model_parallel(rpc_backend=RpcBackend.PROCESS_GROUP,
+                                self_name=None,
```
any reason for making `self_name` an optional arg?

satgera replied:
Thanks for pointing out, this is what caused the test failures. Reverting this.

mrshenli reviewed (commented):
Looking good to me. 

Test failures are real:

```
Sep 09 07:34:24 test_rpc_complex_args (__main__.TestDistAutograd) ... ERROR:root:Caught exception: name 'backend' is not defined, exiting process with exit code: 10
Sep 09 07:34:24 ERROR:root:Caught exception: name 'backend' is not defined, exiting process with exit code: 10
Sep 09 07:34:24 ERROR:root:Caught exception: name 'backend' is not defined, exiting process with exit code: 10
Sep 09 07:34:24 ERROR:root:Caught exception: name 'backend' is not defined, exiting process with exit code: 10
Sep 09 07:34:24 FAIL
```

satgera committed:
Update on "Dynamic registration of RPC backends"


Allow non-pg rpc backends to be plugged in as a backend.

Differential Revision: [D17183789](https://our.internmc.facebook.com/intern/diff/D17183789/)

satgera committed:
Update on "Dynamic registration of RPC backends"


Allow non-pg rpc backends to be plugged in as a backend.

Differential Revision: [D17183789](https://our.internmc.facebook.com/intern/diff/D17183789/)

satgera committed:
Update on "Dynamic registration of RPC backends"


Allow non-pg rpc backends to be plugged in as a backend.

Differential Revision: [D17183789](https://our.internmc.facebook.com/intern/diff/D17183789/)

mrshenli commented on torch/distributed/rpc.py:
```diff
@@ -51,8 +52,11 @@ class RpcBackend(Enum):
 
-# TODO: add a context manager to wrap _init_rpc and join_rpc
-def _init_rpc(name, backend=RpcBackend.PROCESS_GROUP):
+# TODO: add a context managet to wrap _init_rpc and join_rpc
```
nit: managet -> manager (I think this comes from rebase)

mrshenli reviewed (approved):
LGTM! Thanks!

mrshenli commented:
@satgera Please make sure the test pass though. The following CI error is real:

```
20:18:45 Running test_rpc ... [2019-09-10 20:18:45.105766]
20:18:47 Traceback (most recent call last):
20:18:47   File "test_rpc.py", line 14, in <module>
20:18:47     BACKEND = getenv('RPC_BACKEND', dist.RpcBackend.PROCESS_GROUP)
20:18:47 AttributeError: 'module' object has no attribute 'RpcBackend'
```

satgera committed:
Update on "Dynamic registration of RPC backends"


Allow non-pg rpc backends to be plugged in as a backend.

Differential Revision: [D17183789](https://our.internmc.facebook.com/intern/diff/D17183789/)

pietern commented on torch/distributed/__init__.py:
```diff
@@ -23,7 +23,10 @@ def is_available():
+        def init_model_parallel(self_name,
+                                rpc_backend=RpcBackend.PROCESS_GROUP,
+                                self_rank=-1,
+                                rpc_init_url=None):
```
Now that you're changing it anyway, can we stick to the same naming as `init_process_group` and use `backend` instead of `rpc_backend` (would you ever need multiple?) and `init_method` instead of `rpc_init_url`.

satgera committed:
Update on "Dynamic registration of RPC backends"


Allow non-pg rpc backends to be plugged in as a backend.

Differential Revision: [D17183789](https://our.internmc.facebook.com/intern/diff/D17183789/)
